lottie-swift/src/Private/Model/Animation.swift

```diff
@@ -0,0 +1,101 @@
+//
+//  Composition.swift
+//  lottie-swift
+//
+//  Created by Brandon Withrow on 1/7/19.
+//
+
+import Foundation
+
+public enum CoordinateSpace: Int, Codable {
+  case type2d
+  case type3d
+}
+
+/**
+ The `Animation` model is the top level model object in Lottie.
+ 
+ An `Animation` holds all of the animation data backing a Lottie Animation.
+ Codable, see JSON schema [here](https://github.com/airbnb/lottie-web/tree/master/docs/json).
+ */
+public class Animation: Codable {
+  
+  /// The version of the JSON Schema.
+  let version: String
+  
+  /// The coordinate space of the composition.
+  let type: CoordinateSpace
+  
+  /// The start time of the composition in frameTime.
+  public let startFrame: AnimationFrameTime
+  
+  /// The end time of the composition in frameTime.
+  public let endFrame: AnimationFrameTime
+  
+  /// The frame rate of the composition.
+  public let framerate: Double
+  
+  /// The height of the composition in points.
+  let width: Int
+  
+  /// The width of the composition in points.
+  let height: Int
+  
+  /// The list of animation layers
+  let layers: [LayerModel]
+  
+  /// The list of glyphs used for text rendering
+  let glyphs: [Glyph]?
+  
+  /// The list of fonts used for text rendering
+  let fonts: FontList?
+  
+  /// Asset Library
+  let assetLibrary: AssetLibrary?
+  
+  /// Markers
+  let markers: [Marker]?
+  let markerMap: [String : Marker]?
+  
+  enum CodingKeys : String, CodingKey {
+    case version = "v"
+    case type = "ddd"
+    case startFrame = "ip"
+    case endFrame = "op"
+    case framerate = "fr"
+    case width = "w"
+    case height = "h"
+    case layers = "layers"
+    case glyphs = "chars"
+    case fonts = "fonts"
+    case assetLibrary = "assets"
+    case markers = "markers"
+  }
+  
+  required public init(from decoder: Decoder) throws {
+    let container = try decoder.container(keyedBy: Animation.CodingKeys.self)
+    self.version = try container.decode(String.self, forKey: .version)
+    self.type = try container.decodeIfPresent(CoordinateSpace.self, forKey: .type) ?? .type2d
+    self.startFrame = try container.decode(AnimationFrameTime.self, forKey: .startFrame)
+    self.endFrame = try container.decode(AnimationFrameTime.self, forKey: .endFrame)
+    self.framerate = try container.decode(Double.self, forKey: .framerate)
+    self.width = try container.decode(Int.self, forKey: .width)
+    self.height = try container.decode(Int.self, forKey: .height)
+    self.layers = try container.decode([LayerModel].self, ofFamily: LayerType.self, forKey: .layers)
+    self.glyphs = try container.decodeIfPresent([Glyph].self, forKey: .glyphs)
+    self.fonts = try container.decodeIfPresent(FontList.self, forKey: .fonts)
+    self.assetLibrary = try container.decodeIfPresent(AssetLibrary.self, forKey: .assetLibrary)
+    self.markers = try container.decodeIfPresent([Marker].self, forKey: .markers)
+    
+    if let markers = markers {
+      var markerMap: [String : Marker] = [:]
```

Comment:
Any thoughts on writing this whole if-let-else statement to this?

self.markerMap = markers?.reduce([String : Marker](),  { makerMap,  marker, in
    markerMap[marker.name] = marker
}

Reply:
[`reduce(::)`](https://developer.apple.com/documentation/swift/array/2298686-reduce) will copy the dictionary at each iteration. You should use `reduce(into::)` for this!

Reply:
You’re right, I did mean ‘reduce(into::)’ as my code probably won’t compile.

Reply:
👍